### PR TITLE
ddtrace/tracer: StartSpanFromContext should work nil ctx on Go 1.15

### PR DESCRIPTION
**Pull request by evanj**
Previously, StartSpanFromContext would "work" with nil ctx. Using the
ctx (e.g. ctx.Value) would later cause a segfault. With Go 1.15,
calling StartSpanFromContext now panics. Preserve the existing
behaviour by defaulting to context.Background(). Document this
behaviour. Add a test.

Without the fix, on Go 1.14 this new test panics with "runtime error:
invalid memory address or nil pointer dereference" when calling
context.Value. On Go 1.15 it fails with "cannot create context from
nil parent" when calling StartSpanFromContext.

### CODE REVIEW

File: ddtrace/tracer/context_test.go [hunk 1]
```diff
@@ -74,3 +74,24 @@ func TestStartSpanFromContext(t *testing.T) {
 	assert.Equal("gin", got.Service)
 	assert.Equal("/", got.Resource)
 }
+
+func TestStartSpanFromNilContext(t *testing.T) {
+	_, _, _, stop := startTestTracer(t)
+	defer stop()
+
+	// Go before 1.15: StartSpanFromContext(nil, ...) worked; calling ctx.Value would segfault
+	// Go == 1.15: StartSpanFromContext(nil, ...) would panic: cannot create context from nil parent
```
Comment (knusbaum) [inline ddtrace/tracer/context_test.go:83]: We can leave these comments out. This commit fixes the panic, so explaining in the comment is unnecessary.

**Review by knusbaum — CHANGES_REQUESTED** [review]
👍 
Generally we shouldn't detect and correct `nil` contexts, since [passing nil contexts is invalid](https://godoc.org/context), rather fixing it in the caller.

However, on the API boundary I think it's a good idea. Plus, other existing functions already check for and accept `nil` contexts, like `SpanFromContext`.

**Review by knusbaum — DISMISSED** [review]
LGTM. Thanks.

File: ddtrace/tracer/context.go [hunk 2]
```diff
@@ -38,7 +38,13 @@ func SpanFromContext(ctx context.Context) (Span, bool) {
 // StartSpanFromContext returns a new span with the given operation name and options. If a span
 // is found in the context, it will be used as the parent of the resulting span. If the ChildOf
 // option is passed, the span from context will take precedence over it as the parent span.
+// If ctx is nil, this will use context.Background().
 func StartSpanFromContext(ctx context.Context, operationName string, opts ...StartSpanOption) (Span, context.Context) {
+	// default to context.Background() to avoid panics on Go >= 1.15
```
Comment (gbbr) [inline ddtrace/tracer/context.go:43]: Can you please move this comment inside the `if` statement?

File: ddtrace/tracer/context.go [hunk 3]
```diff
@@ -38,7 +38,13 @@ func SpanFromContext(ctx context.Context) (Span, bool) {
...
+	if ctx == nil {
+		ctx = context.Background()
+	}
+
```
Comment (gbbr) [inline ddtrace/tracer/context.go:47]: And remove this empty line please.

**Review by gbbr — CHANGES_REQUESTED** [review]

**Comment by evanj** [comment]
I have made the suggested changes; thanks!

**Review by knusbaum — DISMISSED** [review]

File: ddtrace/tracer/context.go [hunk 4]
```diff
@@ -38,7 +38,12 @@ func SpanFromContext(ctx context.Context) (Span, bool) {
 // StartSpanFromContext returns a new span with the given operation name and options. If a span
 // is found in the context, it will be used as the parent of the resulting span. If the ChildOf
 // option is passed, the span from context will take precedence over it as the parent span.
+// If ctx is nil, this will use context.Background().
```
Comment (gbbr) [inline ddtrace/tracer/context.go:41]: This is nothing but a safety mechanism, I don't think it warrants documentation changes.

**Review by gbbr — COMMENTED** [review]
Thanks for making the changes, and sorry for coming back with yet another request. Can you please apply it and then we're good to merge? 🙏

**Review by knusbaum — APPROVED** [review]

**Review by gbbr — APPROVED** [review]

**Pull request merged**